pyds9/pyds9.py

```diff
@@ -610,184 +630,270 @@ def access(self):
-                    narr = arr.astype(numpy.float32)
-                else:
-                    narr = arr
-            if not narr.flags['C_CONTIGUOUS']:
-                narr = numpy.ascontiguousarray(narr)
-            bp = _np2bp(narr.dtype)
-            buf = narr.tostring('C')
-            blen = len(buf)
-            (w, h) = narr.shape
-
-            # note that this needs the "endian=" part because sometimes it's
-            # left out completely
-            endianness = ''
-            if narr.dtype.byteorder == '=':
-                endianness = ',endian=' + sys.byteorder
-            elif narr.dtype.byteorder == '<':
-                endianness = ',endian=little'
-            elif narr.dtype.byteorder == '>':
-                endianness = ',endian=big'
-
-            paramlist = 'array [xdim={0},ydim={1},bitpix={2}{3}]'
-            return self.set(paramlist.format(h, w, bp, endianness), buf,
-                            blen+1)
+        Returns
+        -------
+        :class:`six.BytesIO`
+            file-like object containing the FITS data
+        '''
+        self._selftest()
+        imgData = self.get('fits')
+        return BytesIO(string_to_bytes(imgData))
+
+    def _hdulist_to_ds9_fits(self, hdul):
+        '''Send the input HDUList to ds9
+
+        Parameters
+        ----------
+        hdul : :class:`HDUList` (astropy or pyfits)
+            FITS file to send
+
+        Returns
+        -------
+        success : int
+            1 indicates that ds9 was contacted successfully, while a return
+            value of 0 indicates a failure.
+        '''
+        self._selftest()
+        # for python2 BytesIO and StringIO are the same
+        with contextlib.closing(BytesIO()) as newFitsFile:
+            hdul.writeto(newFitsFile)
+            newfits = newFitsFile.getvalue()
+            success = self.set('fits', newfits, len(newfits))
+        return success
+
+    def get_fits(self):
+        """Retrieve data from ds9 as an astropy FITS.
+
+        Examples
+        --------
+
+        >>> hdul = d.get_fits()
+        >>> hdul.info()
+        Filename: StringIO
+        No.    Name         Type      Cards   Dimensions   Format
+        0    PRIMARY     PrimaryHDU      24  (1024, 1024)  float32
+        >>> data = hdul[0].data
+        >>> data.shape
+        (1024, 1024)
+
+        Returns
+        -------
+        :class:`astropy.io.fits.HDUList`
+            FITS object
+        """
+        return fits.open(self._ds9_fits_to_bytes())
 
-    else:
-        def get_arr2np(self):
-            """
-            This method is not defined because numpy in not installed.
-            """
-            raise ValueError('get_arr2np not defined (numpy not found)')
-
-        def set_np2arr(self):
-            """
-            This method is not defined because numpy in not installed.
-            """
-            raise ValueError('set_np2arr not defined (numpy not found)')
+    def set_fits(self, hdul):
```

Comment:
send an ``astropy`` HDUList to ds9

Reply:
I just realised that I need to thing about backward compatibility. I'll think something